File: TinyScript/java/src/main/java/paser/ast/Stmt.java
package paser.ast;

import lexer.Token;
import paser.util.ParseException;
import paser.util.PeekTokenIterator;

import java.util.List;

/**
 * @author MFine
 * @version 1.0
 * @date 2021/8/23 0:19
 **/
public class Stmt extends ASTNode {

    protected Stmt(ASTNodeTypes types, String label) {
        super(types, label);
    }

    public static ASTNode parseStmt(ASTNode parent, PeekTokenIterator it) throws ParseException {
        Token token = it.next();
        Token lookahead = it.peek();
        it.putBack();

        if (token.isVariable() && "=".equals(lookahead.getValue())) {
            return AssignStmt.parse(parent, it);
        } else if ("var".equals(token.getValue())) {
            return DeclareStmt.parse(parent, it);
        } else if ("func".equals(token.getValue())) {
            return FunctionDefineStmt.parse(parent, it);
        }else if ("return".equals(token.getValue())){
            return ReturnStmt.parse(parent,it);
        }else if ("if".equals(token.getValue())){
            return IfStmt.parse(parent,it);
        }
        return null;
    }


}

File: TinyScript/js/src/main/js/parser/ast/Stmt.ts
import ASTNode from "./ASTNode";
import ASTNodeTypes from "./ASTNodeTypes";
import PeekTokenIterator from "../PeekTokenIterator";
import ParseException from "../ParseException";

export default class Stmt extends ASTNode {

    constructor(types: ASTNodeTypes | null, label: string | null) {
        super(types, label);
    }


    static parse(it: PeekTokenIterator): Stmt | null {
        const {AssignStmt, DeclareStmt, IfStmt, Block, FunctionDefineStmt, ReturnStmt,Expr} = require('./index');
        if (!it.hasNext()) {
            throw new ParseException("unexpected token");
        }
        const tkn = it.next();
        const lookahead = it.peek();
        it.putBack();
        if (tkn?.isVariable() && lookahead?.value === '=') {
            return AssignStmt.parse(it);
        } else if (tkn?.value === "var") {
            return DeclareStmt.parse(it);
        } else if (tkn?.value === "if") {
            return IfStmt.parse(it);
        } else if (tkn?.value === "{") {
            return Block.parse(it);
        } else if (tkn?.value === "func") {
            return FunctionDefineStmt.parse(it);
        } else if (tkn?.value === "return") {
            return ReturnStmt.parse(it);
        } else {
            return Expr.parse(it);
        }
    }
}